staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go

```diff
@@ -49,11 +61,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
 			})
 		}
 
-		pod.Spec.HostIPC = true
-		pod.Spec.HostNetwork = true
-		pod.Spec.HostPID = true
+		setHostNamespace(pod, true)
 		return nil
 	default:
 		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileLegacy, target)
 	}
 }
+
+type debugStyle int
+
+const (
+	styleEphemeral debugStyle = iota
+	stylePodCopy
+	styleNode
+)
+
+func getDebugStyle(pod *corev1.Pod, target runtime.Object) debugStyle {
+	switch target.(type) {
+	case *corev1.Pod:
+		if asserted, ok := target.(*corev1.Pod); ok {
+			if pod != asserted { // comparing addresses
+				return stylePodCopy
+			}
+		}
+		return styleEphemeral
+	case *corev1.Node:
+		return styleNode
+	}
+	return debugStyle(-1) // unknown
+}
+
+func (p *generalProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	switch getDebugStyle(pod, target) {
+	case stylePodCopy:
+		// For copy of pod: sets SYS_PTRACE in debugging container, sets shareProcessNamespace
+		pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+		modifyContainer(pod.Spec.Containers, containerName, func(c *corev1.Container) {
+			c.SecurityContext = addCapability(c.SecurityContext, "SYS_PTRACE")
+		})
+		removeProbes(pod.Spec.Containers)
+
+	case styleEphemeral:
+		// For ephemeral container: sets SYS_PTRACE in ephemeral container
+		modifyEphemeralContainer(pod.Spec.EphemeralContainers, containerName, func(c *corev1.EphemeralContainer) {
+			c.SecurityContext = addCapability(c.SecurityContext, "SYS_PTRACE")
+		})
+
+	case styleNode:
+		// empty securityContext; uses host namespaces, mounts root partition
+		const volumeName = "host-root"
+		pod.Spec.Volumes = append(pod.Spec.Volumes, corev1.Volume{
+			Name: volumeName,
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{Path: "/"},
+			},
+		})
+		modifyContainer(pod.Spec.Containers, containerName, func(c *corev1.Container) {
+			c.VolumeMounts = append(c.VolumeMounts, corev1.VolumeMount{
+				MountPath: "/host",
+				Name:      volumeName,
+			})
+		})
+		pod.Spec.SecurityContext = nil
+		setHostNamespace(pod, true)
+	default:
+		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileGeneral, target)
+	}
+
+	return nil
+}
+
+func (p *baselineProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	switch getDebugStyle(pod, target) {
+	case stylePodCopy:
+		// For copy of pod: empty securityContext; sets shareProcessNamespace
+		pod.Spec.SecurityContext = nil
+		pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+		setHostNamespace(pod, false)
```

Comment:
I agree that all these functions `setHostNamespace`, `modifyContainer`, `removeProbes` preventing a lot of duplications and it causes less code and appreciate the effort. However, I think duplication is not a bad thing in this case. Because when someone in the future read the code, it immediately should understand what all these profiles are doing.

For example, what we are losing by this;

```go
// For copy of pod: empty securityContext; sets shareProcessNamespace
pod.Spec.SecurityContext = nil
pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
pod.Spec.HostNetwork = false
pod.Spec.HostPID = false
pod.Spec.HostIPC = false
for i, c := range pod.Spec.Containers {
   pod.Spec.Containers[i].SecurityContext =  &corev1.SecurityContext{
			Capabilities: &corev1.Capabilities{
				Add: []corev1.Capability{c},
			},
		}
  pod.Spec.Containers[i].LivenessProbe = nil
  pod.Spec.Containers[i].ReadinessProbe = nil
}

```

Reply:
imo, for profiles we want to prioritize composability. We have a lot of profiles that differ in only minor ways. From a readability perspective, I think what's interesting is not what each profile does (which is simple) but how it differs from the other profiles (which is complicated). Helpers make it easy to see how profiles differ.

I generally agree that duplication is not necessarily a bad thing, and it's not duplication that I'm trying to avoid. It was my hope that careful helper naming would improve readability, like:

```
general:
  RemoveLabelsAndProbes()
  case styleNode:
    ClearSecurityContext()
    UseHostNamespaces()
    MountRootPartition()
  case stylePodCopy:
    AllowProcessTracing()
    ShareProcessNamespace()
  case styleEphemeral:
    AllowProcessTracing()

baseline:
  RemoveLabelsAndProbes()
  case styleNode:
    ClearSecurityContext()
  case stylePodCopy:
    ClearSecurityContext()
    ShareProcessNamespace()
  case styleEphemeral:
    ClearSecurityContext()

restricted:
  DisallowRoot()
  DropCapabilities()
  case stylePodCopy:
    ShareProcessNamespace()
  case styleEphemeral:
    ClearSecurityContext()
```

What do you think?

Reply:
I think you are right, in order to increase readability further, with good naming we can apply your suggestion

Reply:
Good idea, implemented in 4472c8f6fbdb5d3357fdbaedfb6fcd235a701732